clarify text for multiplicative decrease along with adding informative reference

goelvidhi opened the pull request:
Closes #102

goelvidhi commented:
I will add issue details to the appendix after #121 is merged.

lisongxu reviewed (changes requested):
"the sender MAY reduce the congestion window immediately upon entering the loss recovery or use other mechanisms, such as Proportional Rate Reduction {{?PRR=RFC6937}}"

We also add a RFC for " immediately upon entering the loss recovery"? Thanks

goelvidhi commented:
> "the sender MAY reduce the congestion window immediately upon entering the loss recovery or use other mechanisms, such as Proportional Rate Reduction {{?PRR=RFC6937}}"
> 
> We also add a RFC for " immediately upon entering the loss recovery"? Thanks

Thats what CUBIC is doing right. CUBIC says,
`When a congestion event is detected by mechanisms described in Section 3.1, CUBIC updates Wmax and reduces cwnd and ssthresh immediately as described below`

So, either do it immediately what CUBIC says or use something else like PRR.

goelvidhi commented:
@lisongxu maybe I misunderstood what you said. Are you saying we should add an RFC when we say `reduce immediately`? What would you suggest? Also, please see my other comment.

lisongxu commented:
@vidhigoel-apple Yes, I was thinking to add a related RFC when we say "reduce cwnd immediately". I am not sure whether this is an accurate description, as Reno and others set cwnd to ssthresh at the end of fast recovery instead of immediately.

nealcardwell commented:
Adding a related RFC to document the "reduce cwnd immediately" alternative sounds good. I think [RFC6675] would be appropriate, since both specifies reducing cwnd immediately upon entering recovery:

RFC5681 at
  https://datatracker.ietf.org/doc/html/rfc5681#section-3.2
says:
   2.  When the third duplicate ACK is received, a TCP MUST set ssthresh
       to no more than the value given in equation (4).
...
   3.  The lost segment starting at SND.UNA MUST be retransmitted and
       cwnd set to ssthresh plus 3*SMSS. 

In RFC6675 at
  https://datatracker.ietf.org/doc/html/rfc6675#section-5
it says:

   (1) If DupAcks >= DupThresh, go to step (4).
...
 (4) Invoke fast retransmit and enter loss recovery as follows:
...
       (4.2) ssthresh = cwnd = (FlightSize / 2)

             The congestion window (cwnd) and slow start threshold
             (ssthresh) are reduced to half of FlightSize per [RFC5681].
             Additionally, note that [RFC5681] requires that any
             segments sent as part of the Limited Transmit mechanism not
             be counted in FlightSize for the purpose of the above
             equation.

       (4.3) Retransmit the first data segment presumed dropped ...

lisongxu commented:
Thank you all for the discussion! This is very helpful.

Based on the discussions, it seems that it is better to clearly indicate that Cubic reduces ssthresh at the beginning of fast recovery, and also updates W_max (fast convergence) at the beginning of the fast recovery.  Cubic may reduce cwnd at the beginning or the end of fast recovery depending on the specific recovery algorithm.

goelvidhi commented:
I have addressed the comments to include everyone's points. Please see the new changes.

@markkukojo I would like to keep the recommendation for PRR as it provides guidance for alternate approaches that may be better than what is described in CUBIC.

goelvidhi commented:
@markkukojo @nealcardwell @lisongxu @larseggert please review again.
